src/Utils/Logging.hs

```diff
@@ -252,90 +263,183 @@ genericLogHandle f b msg = case fromBackendLogMessage msg of
 -- default level of the filter.
 --
--- These semantics seem to be useful under certain circumstances. At least, the
--- order of the rules doesn't matter.
---
 -- When a filter is specified more than once in a configuration, filters are
--- merged by concatenating the lists of rules and by taking the minimum log
--- level.
+-- merged by concatenating the lists of rules (including default rules)
 --
--- The default log filter has no rules and log level debug as default level.
+-- The default log filter has no rules and log level debug as default level
+-- and 1 as default rate.
 --
 data LogFilter = LogFilter
-    { _logFilterRules :: ![(L.LogLabel, L.LogLevel)]
-    , _logFilterDefault :: !L.LogLevel
+    { _logFilterRules :: ![LogFilterRule]
+    , _logFilterDefaultLevel :: !L.LogLevel
+    , _logFilterDefaultRate :: !Probability
     }
     deriving (Show, Eq, Ord, Generic)
 
 makeLenses ''LogFilter
 
+propMult :: Probability -> Probability -> Probability
+propMult (Probability a) (Probability b) = Probability (a * b)
+
 instance Semigroup LogFilter where
     a <> b = LogFilter
         { _logFilterRules = _logFilterRules a <> _logFilterRules b
-        , _logFilterDefault = min (_logFilterDefault a) (_logFilterDefault b)
+        , _logFilterDefaultLevel = min (_logFilterDefaultLevel a) (_logFilterDefaultLevel b)
+        , _logFilterDefaultRate =
+            case compare (_logFilterDefaultLevel a) (_logFilterDefaultLevel b) of
+                LT -> _logFilterDefaultRate a
+                EQ -> propMult (_logFilterDefaultRate a) (_logFilterDefaultRate b)
+                GT -> _logFilterDefaultRate b
+
         }
     {-# INLINE (<>) #-}
 
 instance Monoid LogFilter where
     mempty = LogFilter
         { _logFilterRules = mempty
-        , _logFilterDefault = maxBound
+        , _logFilterDefaultLevel = maxBound
+        , _logFilterDefaultRate = Probability 1
         }
     {-# INLINE mempty #-}
 
 instance ToJSON LogFilter where
     toJSON a = object
-        [ "rules" .= (f <$> _logFilterRules a)
-        , "default" .=  _logFilterDefault a
+        [ "rules" .= _logFilterRules a
+        , "default" .=  _logFilterDefaultLevel a
+        , "default-rate" .=  _logFilterDefaultRate a
         ]
-      where
-        f ((key, val), lev) = object
-            [ "key" .= key
-            , "value" .= val
-            , "level" .= lev
-            ]
     {-# INLINE toJSON #-}
 
 instance FromJSON LogFilter where
     parseJSON = withObject "LogFilter" $ \o -> LogFilter
-        <$> (o .: "rules" >>= traverse f)
+        <$> o .: "rules"
         <*> o .: "default"
-      where
-        f = withObject "LogRule" $ \o -> (\x y z -> ((x,y),z))
-            <$> o .: "key"
-            <*> o .: "value"
-            <*> o .: "level"
+        <*> o .:? "default-rate" .!= Probability 1
     {-# INLINE parseJSON #-}
 
+logRuleRng :: IORef Prob.Seed
+logRuleRng = unsafePerformIO (Prob.createSystemSeed >>= newIORef)
```

Comment:
Why can't this simply be in `IO`, since it is only used in one place, in IO?

Reply:
`Prob.createSystemSeed` is an expensive operation, in particular when performed on high rate, because it draws entropy from the systems random source.

`logRuleRng` initializes a global `RNG` with a random seed from the system entropy and reuses it after that.

I guess, the RNG could also be initialized higher up in the code and passed down to the use site. It seemed easier to keep things local.

Reply:
How does this code prevent duplicate calls to `createSystemSeed`?

Reply:
Since it is a pure top-level function, it is evaluated only once into a globally shared heap reference (CAF). The `NOINLINE` pragma prevents the code from being inlined and evaluated multiple times.

Reply:
`logRuleRng` is a pure top-level function (CAF), which get evaluated only once into a heap reference. `NOINLINE` prevents the code from getting evaluated repeatedly through inlining.

Reply:
Sounds fine then. Please document the function to this effect, including the importance of NOINLINE